x-pack/plugins/apm/server/lib/services/get_services/get_services_items_stats.ts

```diff
@@ -0,0 +1,309 @@
+import {
+  ServicesItemsSetup,
+  ServicesItemsProjection,
+} from './get_services_items';
+
+const MAX_NUMBER_OF_SERVICES = 500;
+
+const getDeltaAsMinutes = (setup: ServicesItemsSetup) =>
+  (setup.end - setup.start) / 1000 / 60;
+
+interface AggregationParams {
+  setup: ServicesItemsSetup;
+  projection: ServicesItemsProjection;
+}
+
+export const getTransactionDurationAverages = async ({
+  setup,
+  projection,
+}: AggregationParams) => {
+  const { client, indices } = setup;
+
+  const response = await client.search(
+    mergeProjection(projection, {
+      size: 0,
+      index: indices['apm_oss.transactionIndices'],
+      body: {
+        query: {
+          bool: {
+            filter: projection.body.query.bool.filter.concat({
+              term: {
+                [PROCESSOR_EVENT]: ProcessorEvent.transaction,
+              },
+            }),
+          },
+        },
+        aggs: {
+          services: {
+            terms: {
+              ...projection.body.aggs.services.terms,
+              size: MAX_NUMBER_OF_SERVICES,
+            },
+            aggs: {
+              average: {
+                avg: {
+                  field: TRANSACTION_DURATION,
+                },
+              },
+            },
+          },
+        },
+      },
+    })
+  );
+
+  const { aggregations } = response;
+
+  if (!aggregations) {
+    return [];
+  }
+
+  return aggregations.services.buckets.map((bucket) => ({
+    serviceName: bucket.key as string,
+    avgResponseTime: bucket.average.value,
+  }));
+};
+
+export const getAgentNames = async ({
+  setup,
+  projection,
+}: AggregationParams) => {
+  const { client, indices } = setup;
+  const response = await client.search(
+    mergeProjection(projection, {
+      index: [
+        indices['apm_oss.metricsIndices'],
+        indices['apm_oss.errorIndices'],
+        indices['apm_oss.transactionIndices'],
+      ],
+      body: {
+        size: 0,
+        query: {
+          bool: {
+            filter: [
+              ...projection.body.query.bool.filter,
+              {
+                terms: {
+                  [PROCESSOR_EVENT]: [
+                    ProcessorEvent.metric,
+                    ProcessorEvent.error,
+                    ProcessorEvent.transaction,
+                  ],
+                },
+              },
+            ],
+          },
+        },
+        aggs: {
+          services: {
+            terms: {
+              ...projection.body.aggs.services.terms,
+              size: MAX_NUMBER_OF_SERVICES,
+            },
+            aggs: {
+              agent_name: {
+                top_hits: {
+                  _source: [AGENT_NAME],
+                  size: 1,
+                },
+              },
+            },
+          },
+        },
+      },
+    })
+  );
+
+  const { aggregations } = response;
+
+  if (!aggregations) {
+    return [];
+  }
+
+  return aggregations.services.buckets.map((bucket) => ({
```

Comment:
I've seen this in a few places and would be really nice if we could do something like:
```ts
const buckets = aggregations.services.buckets as Buckets<string, {
      agent: {
        name: string;
      };
    }>
```

Reply:
Alternatively:
```ts
type Bucket = <string, { agent: { name: string; }; }>;

return buckets.map(bucket: Bucket => {
 serviceName: bucket.key,
 agentName: bucket.agent_name.hits.hits[0]?._source.agent.name
})
```

Reply:
I'm not sure if it's that easy, but I might be missing something. You'd also need to define `agent_name`, the name of the sub-aggregation, and the fact that it's a `top_hits` aggregation. Let me see if we still need it anyway if we resolve the document types from `apm.events`.

Reply:
Looks like we don't:

![image](https://user-images.githubusercontent.com/352732/86402474-2935a600-bcac-11ea-9d6a-f9ca50a95807.png)

Reply:
Beautiful! Any way to narrow the source down even more? I seem to recall we could do that with the client's first generic argument: `ApmEventsClient<Transaction>()`

Reply:
it's based on whatever you specify in `apm.events`. In this case, it searches on error, metric and transaction indices, that's why it's a union type. If it were just `ProcessorEvent.transaction`, the type would be `Transaction`.

Reply:
Oh, wow. I guess that's correct then 👍